docs/dev/relay_pass_infra.rst

```diff
@@ -0,0 +1,616 @@
+    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+    KIND, either express or implied.  See the License for the
+    specific language governing permissions and limitations
+    under the License.
+
+Relay Pass Infra
+==================================
+
+Relay features a series of optimization passes which improve performance metrics
+of models such as mean inference, memory footprint, or power consumption for
+specific devices. There are a suite of standard optimizations as well as machine
+learning specific optimizations including constant folding, dead code
+elimination, operator layout alteration, and operator fusion, etc. Each of these
+passes are structured as Relay-to-Relay transformations on the abstract syntax
+tree (AST) using the analysis result collected during and/or before traversal.
+
+On one hand, many existing production compilers, such as GCC and LLVM, employ
+pass managers to effectively manage the execution of passes. Initially managing
+passes is straightforward as the number of passes is small, but mature compilers
+will contain hundreds of individual passes. Often external users will want to
+have custom passes correctly scheduled without having to modify a single
+handcrafted pass order.
+
+On the other hand, modern deep learning frameworks, such as Pytorch and MXNet
+Gluon, also have the tendency to enable pass style layer construction
+scheme through `Sequential`_ and `Block`_, respectively. With such constructs,
+these modern frameworks are able to conveniently add modules/layers to their
+containers and build up neural network easily.
+
+The design of the Relay pass infra is largely inspired by the the hierarchical
+pass manager used in LLVM and the block style containers used in the popular
+deep learning frameworks. The major goals of the pass infra include:
+
+#) enabling better programmatic orchestration of optimizations. This allows
+   users to flexibly customize and build their own optimization pipelines.
+
+#) providing a user friendly way to debug optimization passes.
+
+#) alleviating developers from manually and respectively resolving the
+   dependencies between passes.
+
+#) simplifying the implementation of new passes for developers. For example, we
+   allow users to implement a pass in Python and let the pass infra manipulate
+   its execution.
+
+The Design
+----------
+
+We focus on ease of extension for users, making it possible for users to quickly
+add new passes without loss of backward compatibility. The design contains both
+the backend and the frontend. The former implements the main logic of the pass
+infra. The latter provides simple APIs for users to interact with, i.e.
+allowing users to quickly create their own optimization pipelines.
+
+C++ Backend
+~~~~~~~~~~~
+
+We provide a ``PassInfo`` object to contain the basic information needed by
+a pass. ``name`` is the pass name, ``opt_level`` indicates at which optimization
```

Comment:
Could there be a brief explanation of the opt_level mechanism? Or is it something entirely handled by the pass manager?